File: 백준/Silver III/12789. 도키도키 간식드리미/도키도키 간식드리미.py
n=int(input())
l=list(map(int,input().split()))
m=1
x=[]
while(m!=n+1):  
  if(len(l)!=0):
    if(m!=l[0]):
      if(len(x)!=0 and m==x[0]):
        x.pop(0)
        m+=1
      else:
        x.insert(0,l[0])
        l.pop(0)
    else:
      l.pop(0)
      m+=1
  else:
    if(len(x)!=0):
      if(m!=x[0]):
        break
      else:
        x.pop(0)
        m+=1
if(len(x)==0 and len(l)==0):
  print("Nice")
else:
  print("Sad")

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java
import java.io.*;
import java.util.*;

public class Main {
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int[] arr = new int[n];
		Stack<Integer> temp = new Stack<Integer>();
		StringTokenizer st = new StringTokenizer(br.readLine());
		for(int i = 0; i< n; i++) {
			arr[i]= Integer.parseInt(st.nextToken());
		}
		int left = 1;
		int idx = 1;
		temp.push(arr[0]);
		boolean isit = true;
		while(idx< n) {
			while(!temp.isEmpty() && temp.peek()== left) {
					left+=1;
					temp.pop();
				}
			temp.push(arr[idx++]);
		}
		
		while(!temp.isEmpty()) {
			if(temp.pop() != left) {
				isit= false;
				break;
			}else {
				left++;
			}
		}
		System.out.println((isit) ? "Nice": "Sad");
		
	}
}	